components/vision.py

```diff
@@ -67,11 +81,46 @@ def execute(self) -> None:
         self.ping()
         data = self.vision_data_entry.getDoubleArray(None)
         if data is not None:
+            # add clock offset to vision timestamp
             self.vision_data = VisionData(
                 data[0], data[1], data[2], data[3] + self.get_clocks_offset()
             )
-            # add clock offset to vision timestamp
-
+            if not self.vision_data.timestamp == self.last_data_timestamp:
```

Comment:
```suggestion
            if self.vision_data.timestamp != self.last_data_timestamp:
```
Although, this is long enough that I'd suggest an early return for readability.